### ISSUE TITLE
Better WebAssembly integration

### ISSUE THREAD

**Issue by xtuc** [parcel-bundler/parcel#1325]
It seems that the current WebAssembly integration imposes some restriction: https://github.com/parcel-bundler/parcel/issues/647.

# 💬 RFC

In order to provide the best integration we want wasm to be part of the ES module graph, and treated as any other ES module. It has some issues at the moment, you can find more details here https://github.com/WebAssembly/esm-integration (that we can avoid with some transformations).

Parcel don't necessarily need to expose the `importObject` since it can be automatically wired for the user.

I propose you to use https://github.com/xtuc/webassemblyjs (disclaimer: i'm the author) to decode the wasm and apply transformations if needed (remove the start func, runtime, optimization etc).

## 🔦 Context

I'm currently working on the WebAssembly integration in Webpack.

## 💻 Examples

- wasm decoding and introspection: https://github.com/webpack/webpack/blob/master/lib/wasm/WebAssemblyParser.js

- transformation example: https://github.com/webpack/webpack/blob/master/lib/wasm/WebAssemblyGenerator.js

**Comment by KenanSulayman** [comment]
> I propose you to use https://github.com/xtuc/webassemblyjs (disclaimer: i'm the author) to decode the wasm and apply transformations if needed (remove the start func, runtime, optimization etc).

I'd argue against mutations in the bundling process. Toolchains like `wasm-bindgen` (Rust) already process WebAssembly and generate necessary bindings between Javascript and WebAssembly.

The biggest problem with bundling WebAssembly modules is the asynchronous nature of their compilation when they are inlined as base64 and exceed 4KB. It could be solved by restricting the flexibility of how those modules are loaded, reducing the complexity of building such an integration.

Given that you're working on the WebAssembly integration in Webpack, how did you approach it? And what did you settle on?

**Comment by xtuc** [comment]
> I'd argue against mutations in the bundling process [...]

Yes, the less we mutate the module the less can go wrong, but it was required in Webpack to fit into the ESM module system.

> Toolchains like wasm-bindgen (Rust) [...]

It's tight to Rust, I don't expect using it for bundling any time soon.

> The biggest problem with bundling WebAssembly modules is the asynchronous nature of their compilation [...]

That's wrong, compilation can happen synchronously, even if not recommended (ok for service workers).

> [...] when they are inlined

Inlining modules is not recommended because modules tends to be big.

> [...] and exceed 4KB

Some engines have added a limitation for synchronous compilation. That's not an issue if we use `instantiateStreaming`.

**Comment by KenanSulayman** [comment]
Spawning a _Worker_ to execute something synchronously isn't exactly synchronous. _instantiateStreaming_ is also asynchronous because it's returning a Promise.

(For Node.js specifically --) Parcel currently inlines `fs.readFileSync` as `Buffer('<base64>')` which is reasonable because it doesn't change the behaviour of it being a synchronous operation. But it breaks the WebAssembly instantiation if the module is bigger than 4KB. It can't magically rewrite the Buffer into being a ReadableStream that suddenly makes everything asynchronous -- or change the program code so that it suddenly includes an asynchronous WebAssembly loader. That brings me back to "restricting the flexibility of how those modules are loaded."

If you're using `fetch` to load the WebAssembly module there's no point in bundling it other than identifying the respective external resource and - if at all - rename the reference to carry a checksum.

Finally, if you need to mutate a WebAssembly module in order to fit a specific loader, and make that mutation default for the WebAssembly file type, then it would be better not to have it at all. Instead, this should be built as an extension for parcel that handles this specific use-case for users who choose to opt-in to it.

**Comment by xtuc** [comment]
According to https://github.com/parcel-bundler/parcel/blob/9d6061c2ede15af601764ec102389ba8f3f5f790/src/builtins/loaders/browser/wasm-loader.js#L5 and https://github.com/parcel-bundler/parcel/blob/d27274860eba1ea747f36d002bdc5b8b5f1b8cfe/src/builtins/loaders/node/wasm-loader.js#L14, none of them are inline the buffer.

**Comment by KenanSulayman** [comment]
jfyi, this looks interesting: https://github.com/koute/parcel-plugin-cargo-web

**Comment by lygstate** [comment]
Is that possible to getting webassembly embbed as a string literal, not reading from filesystem? or fetch from remote URI.

**Comment by littledan** [comment]
Note, I've been working on the WebAssembly/ESM integration standard, and I'd like to bring alignment between Parcel and the future built-in browser behavior. You can find the proposal at https://github.com/WebAssembly/esm-integration/blob/master/proposals/esm-integration/README.md . I'm not sure if this is the same as Parcel semantics or something different. Any feedback you have would be welcome.

**Comment by littledan** [comment]
@lygstate You can use data URLs to embed WebAssembly as a string literal in the module specifier (but you might hit size limits). You can use CORS to fetch modules cross origin.

**Comment by lygstate** [comment]
@littledan I know this idea, but need parcel have a option support it

**Comment by devongovett** [comment]
I think we'd like to move forward with better WASM integration in Parcel 2. If someone would like to write up an RFC for what we'd need to do exactly and own this feature, that would be much appreciated!

**Comment by xtuc** [comment]
I would need to take a look at Parcel generated code, because we might need to rewrite the Wasm module a bit or add restrictions.

For context, when instantiating a Wasm module, the imported JavaScript values are copied/snapshoted, as opposed to take their references.
That means the JavaScript value must have been evaluated/initialized before the instantiation.

Similar to JavaScript Modules, if you rely on a linking phase and try to instantiate the module that won't work (what webpack does). Rollup on the other hand will inline/hoist the dependecies before the instantions which just works.

**Comment by samvv** [comment]
Given that I don't have a project to work on aside from my own, I'm willing to give it a try. I'll start by implementing a naive approach to see where the issues lie and then come back with an RFC.

**Comment by mischnic** [comment]
I think the most flexible and least non-standard integration would be something like

- `import {a,b,c} from "./x.wasm"` = the WASM ESM spec
- `import {url, instantiate} from "parcel-wasm:./x.wasm"` = to set memory/imports for more advanced usecases

**Comment by Yongle-Fu** [comment]
https://github.com/lgarron/parcel-wasm-window-load-bug/issues/1

**Comment by Yongle-Fu** [comment]
No transformers found for add.wasm.

**Comment by chinoto** [comment]
@mischnic The way the plugin for Vite works is rather nice:
```ts
import init, { Universe, Cell } from "wasm-game-of-life";
let { memory } = await init();
```
^ [Example](https://github.com/chinoto/wasm_game_of_life/blob/master/www/index.ts#L1) from following the WASM Game of Life guide.
